lib/Transforms/Utils/LCSSA.cpp

```diff
@@ -245,6 +246,15 @@ bool llvm::formLCSSA(Loop &L, DominatorTree &DT, LoopInfo *LI,
            !isa<PHINode>(I->user_back())))
         continue;
 
+      // HLSL changes begin
+      // Skip inserting redundant PHI nodes on instruction of struct type
+      // as the resultant IR is not considered a valid DXIL.
```

Comment:
Isn't it a problem that a function that claim to produce LCSSA form may not actually do so? What happens if struct-producing instructions from in the loop are referenced outside of it?

Reply:
Note that the phi node inserted by LCSSA is **redundant** one and is only inserted to have a loop-closed SSA which makes the code shape amenable to other loop optimizations. In other words, even if we don't insert the redundant phi node, it would still be a valid SSA.

Reply:
Valid SSA, yes, but going from its name one could assume that after calling this method you have a valid LCSSA form. The phi node is redundant for SSA but is it required to be valid LCSSA?

Reply:
All right, we talked and I understand why this is ok, though it'd be good if the comment could be say why DXIL doesn't like phis of structs in this case.